packages/template-retail-react-app/app/utils/routes-utils.js

```diff
@@ -9,25 +9,99 @@ import {getSites} from '@salesforce/retail-react-app/app/utils/site-utils'
 import {urlPartPositions} from '@salesforce/retail-react-app/app/constants'
 
 /**
- * Construct literal routes based on url config
- *      with site and locale references (ids and aliases) from each in your application config
+ * Build regex patterns for site and locale route parameters.
+ * Creates patterns like "siteA|siteB|siteC" from all valid site/locale refs.
+ *
+ * @param {array} allSites - array of site configurations
+ * @returns {object} - { sitePattern, localePattern }
+ */
+const buildRoutePatterns = (allSites) => {
+    const siteRefs = allSites.flatMap((site) => [site.alias, site.id]).filter(Boolean)
+
+    const localeRefs = allSites
+        .flatMap((site) => site.l10n.supportedLocales)
+        .flatMap((locale) => [locale.alias, locale.id])
+        .filter(Boolean)
+
+    // Remove duplicates and join into regex pattern
+    const sitePattern = [...new Set(siteRefs)].join('|')
+    const localePattern = [...new Set(localeRefs)].join('|')
+
+    return {sitePattern, localePattern}
+}
+
+/**
+ * Configure routes using parameterized paths with regex constraints.
+ * This approach generates fewer routes by using patterns like:
+ *   /:site(siteA|siteB)/:locale(en|fr)/path
+ *
+ * Note: This may match site/locale combinations that aren't valid together
```

Comment:
Whats the strategy for runtime validation? Perhaps just a simple hook at the start of all routes to validate the site/locale combo and 404 otherwise?